pkg/download/download.go

```diff
@@ -67,43 +71,94 @@ func Download(endpoint string, opts ...Option) (b []byte, err error) {
 		opt(dlOpts)
 	}
 
-	req, err := http.NewRequest(http.MethodGet, u.String(), nil)
-	if err != nil {
-		return b, err
-	}
+	switch u.Scheme {
+	case "http", "https":
+		var req *http.Request
 
-	for k, v := range dlOpts.Headers {
-		req.Header.Set(k, v)
-	}
+		if req, err = http.NewRequest(http.MethodGet, u.String(), nil); err != nil {
+			return b, err
+		}
 
-	err = retry.Exponential(60*time.Second, retry.WithUnits(time.Second), retry.WithJitter(time.Second)).Retry(func() error {
-		b, err = download(req)
-		if err != nil {
-			return retry.ExpectedError(err)
+		for k, v := range dlOpts.Headers {
+			req.Header.Set(k, v)
 		}
 
-		if dlOpts.Format == b64 {
-			var b64 []byte
+		err = retry.Exponential(60*time.Second, retry.WithUnits(time.Second), retry.WithJitter(time.Second)).Retry(func() error {
+			b, err = handleHTTP(req)
+			if err != nil {
+				return retry.ExpectedError(err)
+			}
 
-			b64, err = base64.StdEncoding.DecodeString(string(b))
+			return nil
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to download config from %q: %w", u.String(), err)
+		}
+	case "tftp":
+		addr := u.Hostname()
+		if u.Port() == "" {
+			addr += ":69"
+		}
+
+		err = retry.Exponential(60*time.Second, retry.WithUnits(time.Second), retry.WithJitter(time.Second)).Retry(func() error {
+			b, err = handleTFTP(addr, u.Path)
 			if err != nil {
 				return err
 			}
 
-			b = b64
+			return nil
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to download config from %q: %w", u.String(), err)
 		}
+	default:
+		return nil, fmt.Errorf("unsupported scheme: %q", u.Scheme)
+	}
 
-		return nil
-	})
-	if err != nil {
-		return nil, fmt.Errorf("failed to download config from %q: %w", u.String(), err)
+	if dlOpts.Format == b64 {
+		var b64 []byte
+
+		b64, err = base64.StdEncoding.DecodeString(string(b))
+		if err != nil {
+			return nil, err
+		}
+
+		b = b64
 	}
 
 	return b, nil
 }
 
-// download handles the actual http request
-func download(req *http.Request) (data []byte, err error) {
+func handleTFTP(addr, filename string) (data []byte, err error) {
+	c, err := tftp.NewClient(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := c.Receive(filename, "octet")
+	if err != nil {
+		return nil, err
+	}
+
+	var b bytes.Buffer
+
+	written, err := w.WriteTo(bufio.NewWriter(&b))
```

Comment:
I guess wrapping `bytes.Buffer` into `bufio` is kind of overkill, as it's buffering writes to a buffer ;)

afaik `io.Copy` can detect availability of `WriteTo` and `ReadFrom` methods, so `io.Copy(&b, w)` should work as well

`w` should be `r` probably? as it's more of a `Reader`?

Reply:
Awesome. Will update.

Reply:
hmm looks like `io.Copy(&b, w)` won't work because `c.Receive` returns an `io.WriterTo`.

Reply:
Ok fixed. PTAL.